### DISCUSSION TITLE
%+% global data switch

### DISCUSSION THREAD

**Discussion by EvaMaeRey** [ggplot2-extenders/ggplot-extension-club#29] [Q&A]
Just wonder about this operator for new data sometimes.  Seems ggplot2 is so verbal and then there's this guy.  Is it for technical reasons?  Just curious!

**Comment by teunbrand** [comment]
Yes it is for technical reasons. The operators are a bit weird in terms of how OOP (S3/S4) is implemented on both the left-hand as well as the right-hand side, I think. There is already behaviour on how `x + data.frame()` should behave in base R, so ggplot2 cannot override instructions that `<ggplot_object> + data.frame()` means 'store this layer in a particular place in the ggplot object'. The available-but-similar solution is to use `%+%` instead. I suppose it should in theory be possible to get this done properly in S4 or S7, but I haven't explored this thoroughly.

**Reply by EvaMaeRey** [reply]
Ah, interesting.  But, I was wondering about a + new_data(data = my_new_df) alternative.

**Reply by teunbrand** [reply]
An alternative to solve what issue? To get around the awkwardness of `%+%`?

**Reply by EvaMaeRey** [reply]
yup. Just seems like a regular function would be more in step with the syntax. Changing the data is kind of a big move on the one hand, but maybe it's just also just one of a lot of orthogonal big decisions (mapping, facet) where some plot re-specification is possible.

**Reply by teunbrand** [reply]
If you have `new_data()` output an object with a custom class and then write a `ggplot_add()` method for it, I'm sure you could implement this.

**Reply by EvaMaeRey** [reply]
I will look into add method.  I did this object modification way, but I was really more interested + new_data()  

``` r
library(tidyverse)

ggplot(cars) + 
  aes(dist, speed) + 
  geom_point(size = 7) + 
  aes(color = speed) + 
  scale_color_viridis_c(limits = c(0,26)) + 
  scale_x_continuous(limits = c(0,125)) +
  scale_y_continuous(limits = c(0,25))
```

![](https://i.imgur.com/2Rx5Qjn.png)

``` r



plot_update_data <- function(p = last_plot(), data){
  
  p$data <- data

  p

}


plot_update_data(data = cars %>% filter(speed > 10))
```

![](https://i.imgur.com/tzrrI7G.png)

``` r


plot_filter_data <- function(p = last_plot(), keep){
  
  p$data <- p$data %>% filter({{keep}})
  
  p
  
}


plot_filter_data(keep = dist > 40)
```

![](https://i.imgur.com/5IOidQm.png)

<sup>Created on 2024-07-10 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

**Reply by teunbrand** [reply]
Yeah the `ggplot_add()` methods' body should then be very similar to these update functions you've written

**Reply by EvaMaeRey** [reply]
Great!  Will give it a go sometime - put it in {ggwipe} (currently wipe away layers but with a last_plot_wipe approach - can be challenging to think througth, because of auto scale behavior) Nevertheless, excited to `+ wipe_layers()` (all or specified), wipe_layer_last(), `wipe_data()` wiping off data and replace, or `wipe_obs()` for wiping away rows of data!  I'm hopeful the data and obs will feel more straightforward.

**Reply by EvaMaeRey** [reply]
data_replace(), and data_filter() are the names I ended up favoring instead of wipe_data and wipe_observations, which feels a lot less interesting but less confusing.  

I used @yutannihilation's blogpost, https://yutani.rbind.io/post/2017-11-07-ggplot-add/, as a starting point - gghighlight was great jumping off point having the data filtering that's a part of the highlight strategy.  

data_replace() and data_filter() are written up here: https://evamaerey.github.io/mytidytuesday/2024-07-10-ggnewdata/ggnewdata.html

I also created this last guy 'data_mutate_filter' (which I include below) aimed at zapping small counts.  It could tailored further for the count case which is common for bar plots and is also a problem for other geoms that count like ggcirclepack.  

And the explored data (david robinson's tidytuesday code) supports the direction!   If you're gonna go after in-plot data manipulation, probably should be in the realm of mutate, filter, count.  

``` r
library(tidyverse)

drob_funs <- readr::read_csv('https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2024/2024-07-09/drob_funs.csv')
```

``` r
drob_funs %>% 
  ggplot() + 
  aes(id = paste(funs, pkgs)) + 
  ggcirclepack::geom_circlepack() + 
  ggcirclepack::geom_circlepack_text(aes(label = funs)) + 
  coord_equal() +
  aes(fill = pkgs) + 
  guides(fill = "none")
```

![](https://i.imgur.com/QZ8bRFV.png)


``` r

data_mutate_filter <- function(.value, .by, keep) {
  structure(list(sum_specification = rlang::enquo(.value),
                 by_specification = rlang::enquo(.by),
                 keep_specifiction = rlang::enquo(keep)),
            class = "data_mutatefilter")
  
}

ggplot_add.data_mutatefilter <- function(object, plot, object_name) {

  
  new_data <- dplyr::mutate(plot$data, .value =
                            !! object$sum_specification, 
                            .by = !! object$by_specification) %>% 
    dplyr::filter(!! object$keep_specifiction)
    message("New variable named '.value' created")

  plot$data <- new_data
  plot

  
  }
```

``` r

last_plot() +
  data_mutate_filter(.value = n(),       
                     .by = c(funs, pkgs), 
                     keep = .value >= 200) 
#> New variable named '.value' created
```

![](https://i.imgur.com/bOauyN3.png)


This last problem plot I came across, and I can't necessarily speak to, but I assume it's related to how I set up the geom to be computed panel-wise and not group-wise.  

``` r
last_plot() +
  aes(group = pkgs) 
```

![](https://i.imgur.com/AQBMdjc.png)

<sup>Created on 2024-07-17 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

**Reply by EvaMaeRey** [reply]
Now, just using data_filter(), but adding .by argument from dplyr, so can be more flexibly used - don't think data_mutate_filter is needed probably. 

some examples: https://github.com/EvaMaeRey/ggwipe?tab=readme-ov-file#explore-extension-exported-functions

So, for example data_filter(n()>50, .by = user) to narrow down to some extenders who have delivered a lot of exported function:  
![image](https://github.com/user-attachments/assets/dc049deb-5ac1-4167-b724-985045703a29)

Unsure about the data_* prefix. Maybe df_* is better.

**Comment by EvaMaeRey** [comment]
I'm wondering about the '.' behavior that you can do to retrieve current in layers for the proposed data_replace() function.  

`geom_point(data = . %>% filter(var == "my interesting category", color = "red")`

In the pipeline below, I'd wondered about replacing doubly highlighted part with a dot, i.e. referring to the *current plot data* (which is a filtered to  artist == "Stevie Wonder" instead refiltering in the replace data step (all the way back to `songs` dataframe and refiltering).   But it doesn't work.    I'm guessing that the ggplot_add might be a discontinuity that won't allow for that?   I remember someone being pretty surprised at a meetup, to see that '.' even works in geom_*() context, something to do with lambda.

![image](https://github.com/user-attachments/assets/fec5b499-25e0-422f-8fdb-790519865dd6)

data_replace is defined w/ ggplot_add method:

data_replace <- function(data) {
  structure(list(new_data_specification = data), 
            class = "wipedata")
}

ggplot_add.wipedata <- function(object, plot, object_name) {
  
  plot$data <- object$new_data_specification
  plot

  }

**Reply by teunbrand** [reply]
The {magrittr} piping with `.` returns a function (of the 'functional sequence' subclass), so that kicks off the 'function as `data`' clause of the `data` documention.

You could probably do something like the following to allow that in `data_replace()` too:
```r
ggplot_add.wipedata <- function(object, plot, object_name) {
  new <- object$new_data_specifiction
  if (is.function(new)) {
    new <- new(plot$data)
  }
  plot$data <- new
  plot
}
```

**Reply by EvaMaeRey** [reply]
I'll give it a try!  Thank you for lending your expertise to this experiment.  

data_replace lets you do some in-plot-pipeline data cleaning.  I thought it felt fun - more fun than regular data cleaning - the circles collapse together when categories collapse.  

I also tried to do something like data_mutate(song = str_replace(song, "Bout", "bout")).  That would be nice, but no dice so far.

**Reply by EvaMaeRey** [reply]
I just got around to trying, verbatim, without much luck. Just an initial report. I'm still hopeful!

``` r
library(tidyverse)


# define wipedata method
ggplot_add.wipedata <- function(object, plot, object_name) {
  new <- object$new_data_specifiction
  if (is.function(new)) {
    new <- new(plot$data)
  }
  plot$data <- new
  plot
}

data_replace <- function(data) {
  structure(list(new_data_specification = data), 
            class = "wipedata")
}



circle_packing_data <- data.frame(x = abs(rnorm(100)), id = 1:100) 


# make initial plot
circle_packing_data %>% 
ggplot(data = .) +
  aes(area = x, id = id) + 
  ggcirclepack::geom_circlepack() 
```

![](https://i.imgur.com/0q6t6D4.png)

``` r
# try full replacement w/ fresh data
last_plot() +
  data_replace(data = circle_packing_data %>% mutate(x = sort(x)))
#> Error in panels[[1]]: subscript out of bounds
```

``` r
# try modifying data w/ '.'
last_plot() +
  data_replace(data = . %>% mutate(x = sort(x)))
#> Error in panels[[1]]: subscript out of bounds
```

### modifying method, and referring to the full initial data, back to previous definition we can do the exercise we had in mind which is looking at how circles pack (dependent on id order)

``` r

ggplot_add.wipedata <- function(object, plot, object_name) {

  plot$data <- object$new_data_specification
  plot

  }

last_plot() + 
  data_replace(data = circle_packing_data %>% mutate(x = sort(x)))
```

![](https://i.imgur.com/w99BoCa.png)

``` r


last_plot() + 
  data_replace(data = circle_packing_data %>% mutate(x = rev(sort(x))))
```

![](https://i.imgur.com/N4Kq3g5.png)

``` r
last_plot() %+% 
  (. %>% mutate(x = sort(x)))
#> Error in `ggplot_add()`:
#> ! Can't add `(. %>% mutate(x = sort(x)))` to a <ggplot> object
#> ℹ Did you forget to add parentheses, as in `(. %>% mutate(x = sort(x)))()`?
```

<sup>Created on 2024-08-12 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

**Reply by EvaMaeRey** [reply]
For the circle packing exercise, it will be more illustrative to do something like mutate(id = fct_reorder(id, by = x))
